src/tree/updater_gpu_hist.cu

```diff
@@ -550,16 +515,63 @@ struct DeviceShard {
     hist.Reset();
   }
 
+  DeviceSplitCandidate EvaluateSplit(int nidx,
+                                     const HostDeviceVector<int>& feature_set,
+                                     ValueConstraint value_constraint) {
+    dh::safe_cuda(cudaSetDevice(device_id_));
+    auto d_split_candidates = temp_memory.GetSpan<DeviceSplitCandidate>(feature_set.Size());
+    DeviceNodeStats node(node_sum_gradients[nidx], nidx, param);
+    feature_set.Reshard(GPUSet::Range(device_id_, 1));
```

Comment:
I assume in the future there will be one `feature_set` for each device and the correspondence won't change, otherwise `Reshard` will get us into trouble.

Reply:
Yes we will have to deal with this in the next PR. I am wondering if we can reshard it such that a copy exists on each device.